src/main.ml

```diff
@@ -51,47 +63,52 @@ let require name arg = match arg with
 let require_str name arg =
   require name (if arg = "" then None else Some arg)
 
-let capture_exception f x =
-  Lwt.catch
-    (fun () -> f x >>= fun r -> Lwt.return (`Ok r))
-    (fun e -> Lwt.return (`Error e))
-
-let release_exception = function
-  | `Ok x -> Lwt.return x
-  | `Error e -> Lwt.fail e
-
 let with_block filename f =
   Block.connect filename
   >>= function
-  | `Error _ -> Lwt.fail_with (Printf.sprintf "Unable to read %s" filename)
-  | `Ok x ->
-    capture_exception f x
-    >>= fun r ->
-    Block.disconnect x
-    >>= fun () ->
-    release_exception r
-
-let with_attached_vdi sr vdi read_write f =
-  let pid = Unix.getpid () in
-  let connection_uuid = Uuidm.v `V4 |> Uuidm.to_string in
-  let datapath_id = Printf.sprintf "xapi-nbd/%s/%s/%d" vdi connection_uuid pid in
-  let dbg = Printf.sprintf "xapi-nbd:with_attached_vdi/%s" datapath_id in
-  SM.DP.create ~dbg ~id:datapath_id
-  >>= fun dp ->
-  SM.VDI.attach ~dbg ~dp ~sr ~vdi ~read_write
-  >>= fun attach_info ->
-  SM.VDI.activate ~dbg ~dp ~sr ~vdi
-  >>= fun () ->
-  capture_exception f attach_info.Storage_interface.params
-  >>= fun r ->
-  SM.DP.destroy ~dbg ~dp ~allow_leak:true
-  >>= fun () ->
-  release_exception r
-
-
-let ignore_exn t () = Lwt.catch t (fun _ -> Lwt.return_unit)
-
-let handle_connection xen_api_uri fd tls_role =
+  | `Error e -> Lwt.fail_with (Printf.sprintf "Unable to read %s: %s" filename (Nbd.Block_error_printer.to_string e))
+  | `Ok b ->
+    let block_uuid = Uuidm.v `V4 |> Uuidm.to_string in
+    Hashtbl.add blocks_to_close block_uuid b;
+    Lwt.finalize
+      (fun () -> f b)
+      (fun () ->
+         Block.disconnect b >|= fun () ->
+         Hashtbl.remove blocks_to_close block_uuid
+      )
+
+let with_vbd ~vDI ~vM ~mode ~rpc ~session_id f =
+  Xen_api.VBD.create ~rpc ~session_id ~vM ~vDI ~userdevice:"autodetect" ~bootable:false ~mode ~_type:`Disk ~unpluggable:true ~empty:false ~other_config:[] ~qos_algorithm_type:"" ~qos_algorithm_params:[]
+  >>= fun vbd ->
+  vbds_to_clean_up := StringSet.add vbd !vbds_to_clean_up;
+  Lwt.finalize
+    (fun () ->
+       Lwt_log.notice_f "Plugging VBD %s" vbd >>= fun () ->
+       Xen_api.VBD.plug ~rpc ~session_id ~self:vbd >>= fun () ->
+       Lwt.finalize
+         (fun () -> f vbd)
+         (fun () ->
+            Lwt_log.notice_f "Unplugging VBD %s" vbd >>= fun () ->
+            Xen_api.VBD.unplug ~rpc ~session_id ~self:vbd)
+    )
+    (fun () ->
+       Lwt_log.notice_f "Destroying VBD %s" vbd >>= fun () ->
+       Xen_api.VBD.destroy ~rpc ~session_id ~self:vbd >|= fun () ->
+       vbds_to_clean_up := StringSet.remove vbd !vbds_to_clean_up)
```

Comment:
I don't like accessing references this way, even if it looks safe now. Can you please wrap any change with an `Lwt_mutex.with_lock` (https://ocsigen.org/lwt/2.7.1/api/Lwt_mutex)?

Reply:
Sure, thanks for the suggestion!